lib/phoenix/channel/server.ex

```diff
@@ -10,55 +10,91 @@ defmodule Phoenix.Channel.Server do
+  end
+
   @doc """
   Joins the channel in socket with authentication payload.
   """
-  @spec join(Socket.t, map) :: {:ok, map, pid} | {:error, map}
-  def join(socket, auth_payload) do
-    Phoenix.Endpoint.instrument socket, :phoenix_channel_join,
-      %{params: auth_payload, socket: socket}, fn ->
+  @spec join(Socket.t, module, Message.t, keyword) :: {:ok, map, pid, Socket.t} | {:error, map}
+  def join(socket, channel, message, opts) do
+    %{topic: topic, payload: payload, ref: join_ref} = message
+    assigns = Keyword.get(opts, :assigns, %{})
+
+    socket =
+      %Socket{
+        socket
+        | topic: topic,
+          channel: channel,
+          join_ref: join_ref,
+          assigns: Map.merge(socket.assigns, assigns),
+          private: channel.__socket__(:private)
+      }
+
+    instrument = %{params: payload, socket: socket}
+
+    Phoenix.Endpoint.instrument socket, :phoenix_channel_join, instrument, fn ->
       ref = make_ref()
 
-      case GenServer.start_link(__MODULE__, {socket, auth_payload, self(), ref}) do
+      case Supervisor.start_child(socket.handler, [socket, payload, self(), ref]) do
+        {:ok, :undefined} ->
+          log_join socket, topic, fn -> "Replied #{topic} :error" end
+          receive do: ({^ref, reply} -> {:error, reply})
         {:ok, pid} ->
+          log_join socket, topic, fn -> "Replied #{topic} :ok" end
           receive do: ({^ref, reply} -> {:ok, reply, pid})
-        :ignore ->
-          receive do: ({^ref, reply} -> {:error, reply})
         {:error, reason} ->
           Logger.error fn -> Exception.format_exit(reason) end
           {:error, %{reason: "join crashed"}}
       end
     end
   end
 
+  defp log_join(_, "phoenix" <> _, _func), do: :noop
+  defp log_join(%{private: %{log_join: false}}, _topic, _func), do: :noop
+  defp log_join(%{private: %{log_join: level}}, _topic, func), do: Logger.log(level, func)
+
+  @doc """
+  Gets the socket from the channel.
+  """
+  def socket(pid) do
+    GenServer.call(pid, :socket)
+  end
+
   @doc """
-  Notifies the channel the client closed.
+  Notifies the channels the clients closed.
 
   This event is synchronous as we want to guarantee
-  proper termination of the channel.
+  proper termination of the channels.
   """
-  def close(pid, timeout \\ 5000) do
+  def close(pids, timeout \\ 5000) do
     # We need to guarantee that the channel has been closed
-    # otherwise the link in the transport will trigger it to
-    # crash.
-    ref = Process.monitor(pid)
-    GenServer.cast(pid, :close)
-    receive do
-      {:DOWN, ^ref, _, _, _} -> :ok
-    after
-      timeout ->
-        Process.exit(pid, :kill)
-        receive do
-          {:DOWN, ^ref, _, _, _} -> :ok
-        end
+    # otherwise the link in the transport will trigger it to crash.
+    pids_and_refs =
+      for pid <- pids do
+        ref = Process.monitor(pid)
+        GenServer.cast(pid, :close)
```

Comment:
Interesting - why `cast` instead of `stop`?

Reply:
`stop` would crash if the process is already dead or dies meanwhile.

Reply:
Based on the usage below, the timeout semantics seems weird. If I read this right, it could happen that if you had say three channels which refuse to stop, and a timeout is say 5 seconds, this whole termination code might run for 15 seconds.

Reply:
Good catch! The current code is better than the previous because we at least close them all at the same time but this should be fixed and use a strategy similar to supervisors.

Reply:
Thank you @josevalim - I certainly need to dig that one in a little bit more detail then!